src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,301 @@
+                        // b. Those foreach and 'yield break' (if exists) are last statements in the method (do not count local function declaration statements).
+                        var statementsOnBlockWithForEach = ((BlockSyntax)forEachInfo.ForEachStatement.Parent).Statements
+                            .Where(statement => statement.Kind() != SyntaxKind.LocalFunctionStatement).ToArray();
+                        var lastNonLocalFunctionStatement = statementsOnBlockWithForEach.Last();
+                        if (yieldStatementsCount == 1 && lastNonLocalFunctionStatement == forEachInfo.ForEachStatement)
+                        {
+                            converter = new YieldReturnConverter(forEachInfo, (YieldStatementSyntax)statementCannotBeConverted, yieldBreakStatement: null);
+                            return true;
+                        }
+
+                        // foreach()
+                        // {
+                        //    yield return ...;
+                        // }
+                        // yield break;
+                        // end of member
+                        if (yieldStatementsCount == 2 &&
+                            lastNonLocalFunctionStatement.Kind() == SyntaxKind.YieldBreakStatement &&
+                            !lastNonLocalFunctionStatement.ContainsDirectives &&
+                            statementsOnBlockWithForEach[statementsOnBlockWithForEach.Length - 2] == forEachInfo.ForEachStatement)
+                        {
+                            // This removes the yield break.
+                            converter = new YieldReturnConverter(forEachInfo, (YieldStatementSyntax)statementCannotBeConverted, yieldBreakStatement: (YieldStatementSyntax)lastNonLocalFunctionStatement);
+                            return true;
+                        }
+                    }
+
+                    break;
+            }
+
+            converter = default;
+            return false;
+        }
+
+        protected override SyntaxNode AddLinqUsing(IConverter<ForEachStatementSyntax, StatementSyntax> converter, SemanticModel semanticModel, SyntaxNode root)
+        {
+            var namespaces = semanticModel.GetUsingNamespacesInScope(converter.ForEachInfo.ForEachStatement);
+            if (!(namespaces.Any(namespaceSymbol => namespaceSymbol.Name == "Linq" && namespaceSymbol.ContainingNamespace.Name == "System")) &&
+                 root is CompilationUnitSyntax compilationUnit)
+            {
+                return compilationUnit.AddUsings(SyntaxFactory.UsingDirective(SyntaxFactory.ParseName("System.Linq")));
+            }
+
+            return root;
+        }
+
+        internal static bool IsList(ITypeSymbol typeSymbol, SemanticModel semanticModel)
+            => Equals(typeSymbol?.OriginalDefinition, semanticModel.Compilation.GetTypeByMetadataName(typeof(List<>).FullName));
```

Comment:
nit: your code seems to indicate that typeSymbol could be null.  Consider calling it typeSymbolOpt...